Computed config

davidlehn opened the pull request:
RFC: Update computed config API using a wrapper design.
There are README docs but not yet API docs.
Old API is not yet removed.

mattcollier commented on README.md:
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+// create container object if needed
+c.require('accounts.admin');
+// the config is just a normal object
+config.accounts.admin.name = "Ima Admin";
```
I think this should be `bedrock.config....` for clarity.

dlongley replied:
Single quotes.

mattcollier commented on README.md:
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+c.set('server.domain', 'bedrock.dev');
+// set a computed value that uses values from the main config
+c.setComputed('server.host', () => {
+  return config.server.domain + ':' + config.server.port;
```
again, `bedrock.config...` or define config?

mattcollier commented on README.md:
````diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+```
+
+`setComputed()` can be verbose. a wrapper can be created using the standard
+`bind()` functionality. A helpper called `computer()` will do this for you.
````
typo: helpper

mattcollier commented on README.md:
````diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+Copmuted values can also be created with string templates.
+
+```js
+cc('server.baseUri', 'https://${server.host}');
````
define `cc` again here so the example stands alone?

dlongley commented on README.md:
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
 example, the [bedrock-server][] module's specific configuration object can be
-found under `bedrock.config.server`. Modules may define whatever configuration
+found under `bedrock.config.server`. A `megacorp-feature` module would be found
+under `bedrock['megacorp-feature']`. Modules may define whatever configuration
```
Let's do `mycompany-feature` instead.

dlongley commented on README.md:
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
 use.
 
+The `bedrock.util` module has helpers to setup configurations, and in
+particular, dynamic computed configurations. Computed values can help to
```
dynamically

dlongley commented on README.md:
```diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+values which could be important depending on the use case.)
+
+`bedrock.util.Config` creates a wrapper around a config object and optional
+options. A common setup could look like the following.
```
`This wrapper exposes a new, helpful API that is detailed below. A common...`

dlongley commented on README.md:
````diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+// config is now {"server": {"port": 8443}}
+```
+
+Normal code and the config API can be mixed. A useful helper is `require()`.
````
I actually think `require` is kind of a weird name. `ensure` is another name, but weird for other reasons. Require makes me think you're pulling something in, not making sure it exists.

mattcollier replied:
`init`?

dlongley replied:
`init` would be better too.

davidlehn replied:
I think `init` implies you are fully initializing something vs just making sure it exists.  `init` could be done like set('foo.bar', {})

dlongley commented on README.md:
````diff
@@ -351,10 +351,131 @@ practice for modules to claim their own area in the configuration object is to
+});
+```
+
+Copmuted values can also be created with string templates.
````
Copmuted => Computed

dlongley replied:
`with lodash-style string templates.`

mattcollier replied:
Maybe a link to docs would be helpful.  If we use this syntax, it appears that the lodash website will redirect to the latest API version: `https://lodash.com/docs/#template`

dlongley commented:
There are a few comments here to address, but this is looking really good! Can't wait to be using it everywhere and eliminate some headaches.

davidlehn commented:
Forced update:
- Addressed comments.
- Changed `require()` to `setDefault(path, value)` or `setDefault({path, value, ...})`.
- Removed old API.
- Added README examples to tests.
- Fixed bugs.
